src/engine/safeReader.cpp

```diff
@@ -155,8 +157,53 @@ String SafeReader::readString(size_t stlen) {
 String SafeReader::readString() {
   String ret;
   unsigned char c;
-  while ((c=readC())!=0) {
+  if (isEOF()) throw EndOfFileException(this, len);
+
+  while (!isEOF() && (c=readC())!=0) {
+    ret.push_back(c);
+  }
+  return ret;
+}
+
+String SafeReader::readString_Line() {
+  String ret;
+  unsigned char c;
+  if (isEOF()) throw EndOfFileException(this, len);
+
+  while (!isEOF() && (c = readC()) != 0) {
+    if (c=='\r'||c=='\n') {
+      break;
+    }
+    ret.push_back(c);
+  }
+  return ret;
+}
+
+String SafeReader::readString_Token(unsigned char delim) {
+  String ret;
+  unsigned char c;
+  if (isEOF()) throw EndOfFileException(this, len);
+
+  while (!isEOF() && (c=readC())!=0) {
+    if (c == '\r' || c == '\n') {
+      break;
+    }
+    if (c == delim) {
+      if (ret.length() == 0) {
+        continue;
+      }
+      break;
+    }
     ret.push_back(c);
   }
   return ret;
 }
+String SafeReader::readString_Token() {
+  return readString_Token(' ');
+}
+
+
+
+bool SafeReader::isEOF() {
```

Comment:
Question: is it possible to inline this for performance?

Reply:
Yes!